netket/_vmc.py

```diff
@@ -1,17 +1,100 @@
+import math
 import sys
+import warnings
 
 import numpy as _np
-import math
 
 import netket as _nk
 from netket._core import deprecated
 from .operator import local_values as _local_values
 from netket.stats import statistics as _statistics, mean as _mean
 
-from netket.vmc_common import info
+from netket.vmc_common import info, tree_map
 from netket.abstract_variational_driver import AbstractVariationalDriver
 
 
+# Higher-level VMC functions:
+def estimate_expectations(
+    ops,
+    sampler,
+    n_samples,
+    n_discard=None,
+    compute_gradients=None,
+    return_gradients=False,
+    return_intermediate=False,
+):
+    """
+    For a pytree of linear operators, computes a statistical estimate of the
+    respective expectation values, variances, and optionally gradients of the
+    expectation values with respect to the variational parameters.
+
+    The estimate is based on `n_samples` configurations
+    obtained from `sampler`.
+
+    Args:
+        ops: pytree of linear operators
+        sampler: A NetKet sampler
+        n_samples: Number of MC samples used to estimate expectation values
+        n_discard: Number of MC samples dropped from the start of the
+            chain (burn-in). Defaults to `min(n_samples // 10, 100)`.
+        return_gradients: Whether to compute the gradients of the observables.
+        return_intermediate: Whether to return samples and potentially log-derivatives
+            which have been used to compute the VMC statistics.
+
+    Returns:
+        A pytree of the same structure as `ops` where the leaves are Stats
+        objects containing mean, variance, and MC diagonstics for the corresponding
+        operator. If `return_gradients` is True, the leaves are tuples of
+        MC stats and the gradients, which are ndarray of shape `(psi.n_par,)`.
+
+        If `return_intermediate` is True, the MC samples and log-derivatives
+        are returned as well.
+    """
+    if compute_gradients is not None:
+        warnings.warn(
+            "Argument compute_gradients is deprecated, use return_gradients instead.",
+            FutureWarning,
+        )
+        return_gradients = compute_gradients
```

Comment:
This is an internal API, so it's not covered by semver.
Moreover, we are working on the v3.0 branch anyway.

I would get rid of this deprecation warning and just delete that kwarg

Reply:
The `estimate_expectations` function itself is supposed to be public interface; I put it in `_vmc` for the moment, so that we can remove `variational.py`, but I would still want to expose it somewhere in the future.

But of course, the backwards compatibility can be removed for the 3.0 update, that is true.